include/envoy/network/listen_socket.h

```diff
@@ -90,6 +90,39 @@ class ConnectionSocket {
    * Close the underlying socket.
    */
   virtual void close() PURE;
+
+  /**
+   * Visitor class for setting socket options.
+   */
+  class Options {
+  public:
+    virtual ~Options() {}
+
+    /**
+     * @param socket the socket on which to apply options.
+     * @return true if succeeded, false otherwise.
+     */
+    virtual bool setOptions(ConnectionSocket& socket) const PURE;
+
+    /**
+     * @return bits that can be used to separate upstream connections based on the options. Should
+     *         return zero if connections with different options can be pooled together.
+     */
+    virtual uint32_t hashKey() const PURE;
+  };
+  typedef std::shared_ptr<Options> OptionsSharedPtr;
+
+  /**
+   * Set the socket options to be set on all upstream ClientConnections created to forward
+   * traffic received on this socket.
+   */
+  virtual void setOptionsForUpstreamConnections(const OptionsSharedPtr&) PURE;
```

Comment:
Likewise would prefer to remove the concept of upstream here. Thoughts on removing it?

Reply:
While addressing the comment above I realized that I had not applied the visitor pattern to it's full potential here, see the other response.